Componentes I - TODO list

lmaroso opened the pull request.

Commit:
Update 17-07

Commit:
Update 18-07-19

Commit:
Update 19-07-19

Commit:
Update 22-07-19

Commit:
Update 22-07-19

Commit:
Update 23-07-19

Commit:
Update 24-07-19 working

Commit:
Update 24-07-19 (Refactored)

lmaroso commented:
# Componentes I - TODO list


## Summary

Creation of a basic to-do list app.


## Task functionalities

- _Add new task._
- _Complete a task._
- _Remove a task._
- _Remove completed tasks._

## Screenshots

#### Empty list

![Screenshot_20190724-160526](https://user-images.githubusercontent.com/51789658/61821834-913e0400-ae2e-11e9-8283-3536a5c43fd8.png)

#### Adding three tasks and completing the first one.

![Screenshot_20190724-161158](https://user-images.githubusercontent.com/51789658/61821827-8daa7d00-ae2e-11e9-8653-455f5710bb9e.png)

#### Pressing "Remove selected items" button.


![Screenshot_20190724-161444](https://user-images.githubusercontent.com/51789658/61821841-9438f480-ae2e-11e9-8577-d06f0cb352bc.png)

#### Pressing delete button on the last task.


![Screenshot_20190724-161216](https://user-images.githubusercontent.com/51789658/61821848-9602b800-ae2e-11e9-9362-57e66c8498a1.png)

sdandois commented on package.json:
```diff
@@ -8,8 +8,25 @@
     "lint": "eslint ."
   },
   "dependencies": {
+    "@fortawesome/free-regular-svg-icons": "^5.9.0",
```
Why do you need this dependency?

lmaroso replied:
I don't need it now.
I used it to check if the icons coincided with the ones on the to-do list picture example.
I deleted the library.

sdandois commented on src/Actions/index.js:
```diff
@@ -0,0 +1,38 @@
+    type: ADD_TODO,
+    payload: {
+      text: todo,
+      id: String(generate.next().value),
```
I believe using the generator function to create the ids is a good aproach! But, since the only  purpose is to generate the ids, you could return them already as `strings`. Moreover,  in a design point of view, maybe it would be better that the reducer calculates the `id`.

lmaroso replied:
Fixed!

sdandois commented on src/App/styles.js:
```diff
@@ -0,0 +1,10 @@
+    bottom: 0
+  }
+};
+
```
Redefine this as a  [Stylesheet](https://facebook.github.io/react-native/docs/stylesheet.html).

Use it this way : `styles.footerStyle` (check previous file)

lmaroso replied:
Done!

sdandois commented on src/Components/Buttons/AddButton.js:
```diff
@@ -0,0 +1,21 @@
+import Icon from 'react-native-vector-icons/MaterialIcons';
+import PropTypes from 'prop-types';
+
+import Styles from './Styles';
```
Change both `Styles` to lower case. And use `styles.addButton` (rename it).

lmaroso replied:
Done!

sdandois commented on src/Components/Buttons/AddButton.js:
```diff
@@ -0,0 +1,21 @@
+  const { addButtonStyle } = Styles;
+  return (
+    <TouchableOpacity onPress={onPress} style={addButtonStyle}>
+      <Icon name="add" size={30} color="black" />
```
Extract this number to a constant, named `ICON_SIZE`.

lmaroso replied:
Done!

sdandois commented on src/Components/Buttons/CheckBox.js:
```diff
@@ -0,0 +1,21 @@
+import { View, TouchableOpacity } from 'react-native';
+import PropTypes from 'prop-types';
+
+import Styles from './Styles';
```
Same here with styles

lmaroso replied:
Done!

sdandois commented on src/Components/Buttons/DeleteButton.js:
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
```
This component is _very_ similar to `AddButton`. You could think of a more a general component `IconButton` to use in both cases.

lmaroso replied:
Done!

sdandois commented on src/Components/Buttons/Styles.js:
```diff
@@ -0,0 +1,37 @@
+const Styles = {
```
Changes this to lower case.  Also each component should have it's own directory with it's own styles.js file (with a Stylesheet in it).

 You can group them in Buttons directory but, it is not the way we usually do it . In general, in the projects we have two directories: `components/screens` and` components/common`.

lmaroso replied:
Done!

sdandois commented on src/Components/Footer/Footer.js:
```diff
@@ -0,0 +1,31 @@
+  const { footerStyle } = Styles;
+  return (
+    <View style={footerStyle}>
+      <DeleteSelectedButton onPress={() => dispatchRemoveSelected()} />
```
There is no need of creating the arrow function, you can pass `dispatchRemoveSelected` directly.

Be aware that since you connected this component to redux it is not reusable.

lmaroso replied:
I corrected 'dispatchRemoveSelected' as mentioned.
About the reusability of the component, as it is the footer, I don't need to reuse it.

sdandois commented on src/Components/Input/Input.js:
```diff
@@ -0,0 +1,48 @@
+          style={textInputStyle}
+          placeholder="Enter an Item!"
+          value={text}
+          onChangeText={text => this.setState({ text })}
```
Create a class method(arrow function) with this functionality and call it here.

For example, `onChangeText={this.handleChangeText}`

lmaroso replied:
Done!

sdandois commented on src/Components/Input/Input.js:
```diff
@@ -0,0 +1,48 @@
+import React, { Component } from 'react';
```
Rename this file to `Input/Input.js`. This way you can do the import as `../components/Input` instead of `../components/Input/Input`

Do the same in all components where it applies.

lmaroso replied:
As mentioned, I renamed files that use the same name as there container's folder.
The new name is 'index.js'.

sdandois commented on src/Components/CardSection/Index.js:
```diff
@@ -0,0 +1,17 @@
+/* eslint-disable arrow-body-style */
```
Remove this comment, I believe it is not needed.

lmaroso replied:
Done!

sdandois commented on src/Components/Title/Index.js:
```diff
@@ -0,0 +1,16 @@
+import Styles from './Styles';
+
+const Title = () => {
+  // eslint-disable-next-line no-use-before-define
```
Remove this comment.

lmaroso replied:
Done!

sdandois commented on src/Reducers/index.js:
```diff
@@ -0,0 +1,18 @@
+    case REMOVE_TODO:
+      return state.filter(todo => todo.id !== action.payload);
+    case REMOVE_SELECTED:
+      return state.filter(elem => elem.selected === false);
```
`elem.selected` is already a boolean, you can safely do `elem => !elem.selected)`

lmaroso replied:
Great! I didn't know it works. Thanks for the tip!

sdandois reviewed (changes requested):
Great Work! Check the comments, only small stuff :smile:

Check the linter output also.
```
/home/widergy/work/TrainingLucasMaroso/src/App/index.js
  13:10  error  footerStyle not found in './styles'  import/named

/home/widergy/work/TrainingLucasMaroso/src/Components/List/List.js
  8:17  error  'todos' is missing in props validation  react/prop-types
```

Commit:
Upload files with PR sugested changes by Santi.

lmaroso commented:
## Update:

Add some more fixes that Facu and I detected yesterday:

#### Bug fixed:

- Style import error at "src/App/App.js".
- EsLint error " 'todos' is missing in props validation (react/prop-types)".

#### Modifications:
- Added paddingLeft to Input. It's aligned to the 'paddingLeft' used on the tasklist text.
- Changed footer button's text to "Remove completed tasks".
- Changed "src/Actions/Types.js" file name to "constants.js". This modification is to standardize file names.

#### New features:
- Dismiss keyboard after adding a to-do.

sdandois commented on src/Components/CardSection/styles.js:
```diff
@@ -0,0 +1,20 @@
+    borderBottomWidth: 1,
+    padding: 5,
+    backgroundColor: '#FFF',
+    justifyContent: 'flex-start',
```
You probably not need justifyContent: 'flex-start', as it is the default.

Same for position relative.

sdandois commented on src/Components/CheckBox/index.js:
```diff
@@ -0,0 +1,21 @@
+import styles from './styles';
+
+const CheckBox = ({ selected, onPress }) => {
+  const { checkedCheckBoxStyle, uncheckedCheckBoxStyle } = styles;
```
Do not destructure.

sdandois commented on src/Components/Footer/styles.js:
```diff
@@ -0,0 +1,17 @@
+import { StyleSheet } from 'react-native';
+
+const styles = StyleSheet.create({
+  footerStyle: {
```
Rename this to `footer`. Since you will use like `styles.footer`, there is no need to repeat the word style. Same for textStyle

sdandois commented on src/Components/CheckBox/styles.js:
```diff
@@ -0,0 +1,23 @@
+import { StyleSheet } from 'react-native';
+
+const styles = StyleSheet.create({
+  uncheckedCheckBoxStyle: {
```
Renamed without the  Style sufffix. Check comment above.

sdandois commented on src/Components/CardSection/index.js:
```diff
@@ -0,0 +1,16 @@
+import React from 'react';
```
Change directory name `Components` to lower case `components`.

sdandois reviewed (approved):
Great! Check my comments and take it into account next time.

In the next pr, will have to re-estructure redux actions/reducers.
